lib/libxdp/xsk.c

```diff
@@ -354,22 +364,66 @@ int xsk_umem__create_with_fd(struct xsk_umem **umem_ptr, int fd,
 
 	umem->fill_save = fill;
 	umem->comp_save = comp;
-	*umem_ptr = umem;
-	return 0;
-
+	return umem;
 out_socket:
 	close(umem->fd);
 out_umem_alloc:
 	free(umem);
-	return err;
+err:
+	return libxdp_err_ptr(err, true);
+}
+
+int xsk_umem__create_with_fd(struct xsk_umem **umem_ptr, int fd,
+			     void *umem_area, __u64 size,
+			     struct xsk_ring_prod *fill,
+			     struct xsk_ring_cons *comp,
+			     const struct xsk_umem_config *usr_config)
+{
+	__u32 fill_size, comp_size, frame_size, frame_headroom, flags;
+	struct xsk_umem *umem;
+
+	if(!umem_ptr)
+		return -EFAULT;
+
+	if(!usr_config) {
+		fill_size = XSK_RING_PROD__DEFAULT_NUM_DESCS;
+		comp_size = XSK_RING_CONS__DEFAULT_NUM_DESCS;
+		frame_size = XSK_UMEM__DEFAULT_FRAME_SIZE;
+		frame_headroom = XSK_UMEM__DEFAULT_FRAME_HEADROOM;
+		flags = XSK_UMEM__DEFAULT_FLAGS;
+	} else {
+		/* Set the following fields to their default values if they are passed in as 0 */
+		fill_size = usr_config->fill_size ? usr_config->fill_size : XSK_RING_PROD__DEFAULT_NUM_DESCS;
+		comp_size = usr_config->comp_size ? usr_config->comp_size : XSK_RING_CONS__DEFAULT_NUM_DESCS;
+		frame_size = usr_config->frame_size ? usr_config->frame_size : XSK_UMEM__DEFAULT_FRAME_SIZE;
+		/* The following fields are 0 by default, just keep them as what they are passed in */
+		frame_headroom = usr_config->frame_headroom;
+		flags = usr_config->flags;
+	}
+
+	DECLARE_LIBXDP_OPTS(xsk_umem_opts, opts,
+		.fd = fd,
+		.size = size,
+		.fill_size = fill_size,
+		.comp_size = comp_size,
+		.frame_size = frame_size,
+		.frame_headroom = frame_headroom,
+		.flags = flags,
+	);
```

Comment:
And with that change in `xsk_set_umem_config()` above, this can simply be:
```c
DECLARE_LIBXDP_OPTS(xsk_umem_opts, opts,
		.fd = fd,
		.size = size,
	);
if (usr_config) {
	fill_size = usr_config->fill_size;
	comp_size = usr_config->comp_size;
	frame_size = usr_config->frame_size;
	frame_headroom = usr_config->frame_headroom;
	flags = usr_config->flags;
}
```

Reply:
Cool, this is much more concise